### PR TITLE
example code for line solver

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
this includes
- line solver example
- matching commit for reduced unit tests (Trilinos/trilinos#4444)

### CODE REVIEW

**Comment by kyungjoo-kim** [comment]
spotcheck is not passed yet. I will let you know when it passes.

File: src/batched/KokkosBatched_Copy_Decl.hpp [hunk 1]
```diff
@@ -38,6 +38,32 @@ namespace KokkosBatched {
...
+    struct Copy {
+      template<typename AViewType,
+               typename BViewType>
+      KOKKOS_FORCEINLINE_FUNCTION
+      static int
+      invoke(const MemberType &member,
+             const AViewType &A,
+             const BViewType &B) {
+        int r_val = 0;
+        if (std::is_same<ArgMode,Mode::Serial>::value) {
```
Comment (mhoemmen) [inline src/batched/KokkosBatched_Copy_Decl.hpp]: I feel like this code would be cleaner with a tag to pick serial vs. team, than the current class template specialization design.  I've been trying tag-based design for my in-progress fix for https://github.com/trilinos/Trilinos/issues/4418 and it works really well.  That may be out of scope for this PR, but I think it's worth thinking about for the future.

File: src/batched/KokkosBatched_SetIdentity_Decl.hpp [hunk 2]
```diff
@@ -34,6 +34,28 @@ namespace KokkosBatched {
...
+    template<typename MemberType,
+             typename ArgMode>
+    struct SetIdentity {
+      template<typename AViewType>
+      KOKKOS_FORCEINLINE_FUNCTION
+      static int
+      invoke(const MemberType &member,
+             const AViewType &A) {
+        int r_val = 0;
+        if (std::is_same<ArgMode,Mode::Serial>::value) {
```
Comment (mhoemmen) [inline src/batched/KokkosBatched_SetIdentity_Decl.hpp]: See above.  A tag-based design could let you just write two overloads, instead of an if statement.

Comment (kyungjoo-kim) [reply src/batched/KokkosBatched_SetIdentity_Decl.hpp]: It is all compile time if statement. So, I thought it should not affect the behavior nor performance. However, what you suggested is probably a more clean way to write a code.

File: perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp [hunk 3]
```diff
@@ -0,0 +1,595 @@
...
+    Kokkos::Random_XorShift64_Pool<exec_space> random(13245);
+    Kokkos::fill_random(As, random, value_type(1.0));
+    Kokkos::fill_random(bs, random, value_type(1.0));
+
+    Kokkos::deep_copy(Acopy, As);
+
+    ///
+    /// factorize the matrix
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp]: Why if(1) ?

File: perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp [hunk 4]
```diff
@@ -0,0 +1,595 @@
...
+#if defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOSBATCHED_PROFILE)
+      cudaProfilerStop();
+#endif
+      printf("factorize time = %f , # of factorization per min = %f \n", t, 1.0/t*60);
+    }
+
+    ///
+    /// solve the matrix 20 times
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp]: Same as above.

File: perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp [hunk 5]
```diff
@@ -0,0 +1,595 @@
...
+#if defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOSBATCHED_PROFILE)
+      cudaProfilerStop();
+#endif
+      printf("solve time = %f , # of solves per min = %f\n", t, 1.0/t*60*niter);
+    }
+    
+    ///
+    /// compute residual
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp]: Same as above.

File: perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+
+    /// randomize input
+    Kokkos::Random_XorShift64_Pool<exec_space> random(13245);
+    Kokkos::fill_random(As, random, value_type(1.0));
+    Kokkos::fill_random(bs, random, value_type(1.0));
+
+    ///
+    /// diagonal dominant 
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp]: Same as above

File: perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+      cudaProfilerStop();
+#endif
+    }
+
+    Kokkos::deep_copy(Acopy, As);
+
+    ///
+    /// compute the inverse of diagonals
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp]: I am missing the need for this pattern of If(1) throughout the code.

File: perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+#if defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOSBATCHED_PROFILE)
+      cudaProfilerStop();
+#endif
+      printf("inverse time = %f , # of inverse per min = %f \n", t, 1.0/t*60);
+    }
+
+    ///
+    /// solve the matrix 20 times
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp]: Same as above

File: perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+#if defined(KOKKOS_ENABLE_CUDA) && defined(KOKKOSBATCHED_PROFILE)
+      cudaProfilerStop();
+#endif
+      printf("solve time = %f , # of solves per min = %f\n", t, 1.0/t*60*niter);
+    }
+    
+    ///
+    /// compute residual
+    ///
+    if (1) {
```
Comment (srajama1) [inline perf_test/batched/KokkosBatched_Test_BlockTridiagJacobi.cpp]: Same as above

File: unit_test/CMakeLists.txt [hunk 10]
```diff
@@ -90,16 +90,16 @@ IF (Kokkos_ENABLE_OpenMP)
   
   APPEND_GLOB(OPENMP_BLAS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/openmp/Test_OpenMP_Blas*.cpp)
 
-  IF (KOKKOS_ENABLE_DEBUG)
-    SET(DISABLE_SLOW_DGEMM_DOUBLE_TEST "--gtest_filter=-openmp.gemm_double")
-  ENDIF()
+  # IF (KOKKOS_ENABLE_DEBUG)
+  #   SET(DISABLE_SLOW_DGEMM_DOUBLE_TEST "--gtest_filter=-openmp.gemm_double")
```
Comment (srajama1) [inline unit_test/CMakeLists.txt]: Are we re-enabling this test in debug mode ? Is this planned.

File: unit_test/CMakeLists.txt [hunk 11]
```diff
@@ -146,16 +146,16 @@ IF (Kokkos_ENABLE_Serial)
   
   APPEND_GLOB(SERIAL_BLAS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/serial/Test_Serial_Blas*.cpp)
 
-  IF (KOKKOS_ENABLE_DEBUG)
-    SET(DISABLE_SLOW_DGEMM_DOUBLE_TEST "--gtest_filter=-serial.gemm_double")
-  ENDIF()
+  # IF (KOKKOS_ENABLE_DEBUG)
+  #   SET(DISABLE_SLOW_DGEMM_DOUBLE_TEST "--gtest_filter=-serial.gemm_double")
```
Comment (srajama1) [inline unit_test/CMakeLists.txt]: Same as above.

File: unit_test/blas/Test_Blas1_abs.hpp [hunk 12]
```diff
@@ -153,7 +153,7 @@ int test_abs() {
   Test::impl_test_abs<view_type_a_ll, view_type_b_ll, Device>(0);
   Test::impl_test_abs<view_type_a_ll, view_type_b_ll, Device>(13);
   Test::impl_test_abs<view_type_a_ll, view_type_b_ll, Device>(1024);
-  Test::impl_test_abs<view_type_a_ll, view_type_b_ll, Device>(132231);
+  //Test::impl_test_abs<view_type_a_ll, view_type_b_ll, Device>(132231);
```
Comment (srajama1) [inline unit_test/blas/Test_Blas1_abs.hpp]: I thought this was already pushed as part of another commit.

File: unit_test/blas/Test_Blas1_abs.hpp [hunk 13]
```diff
@@ -162,7 +162,7 @@ int test_abs() {
   Test::impl_test_abs<view_type_a_lr, view_type_b_lr, Device>(0);
   Test::impl_test_abs<view_type_a_lr, view_type_b_lr, Device>(13);
   Test::impl_test_abs<view_type_a_lr, view_type_b_lr, Device>(1024);
-  Test::impl_test_abs<view_type_a_lr, view_type_b_lr, Device>(132231);
+  //Test::impl_test_abs<view_type_a_lr, view_type_b_lr, Device>(132231);
```
Comment (srajama1) [inline unit_test/blas/Test_Blas1_abs.hpp]: Same as above, I won't mark other places below.

File: perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp (hunk 3 above)
Comment (kyungjoo-kim) [reply perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp]: Just my convenience to test and profile the code separately. Profiler returns with weird concatenated names that I cannot figure out which kernels corresponds to which profile output. For convenience, I will keep this if statement.

File: unit_test/CMakeLists.txt (hunk 10 above)
Comment (kyungjoo-kim) [reply unit_test/CMakeLists.txt]: We enable this test from Trilinos after we reduce the test problem sizes. This is matching modification that we already made in Trilinos.

File: unit_test/blas/Test_Blas1_abs.hpp (hunk 12 above)
Comment (kyungjoo-kim) [reply unit_test/blas/Test_Blas1_abs.hpp]: That is on Trilinos. This includes the PR for kokkoskernels.

Comment (srajama1) [reply unit_test/blas/Test_Blas1_abs.hpp]: Got it, Thanks !

File: unit_test/CMakeLists.txt (hunk 10 above)
Comment (srajama1) [reply unit_test/CMakeLists.txt]: Got it, Thanks !

File: perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp (hunk 3 above)
Comment (srajama1) [reply perf_test/batched/KokkosBatched_Test_BlockTridiagDirect.cpp]: This a strange way to profile kernels, but I won't complain :)

**Comment by ndellingwood** [comment]
Merging this in (needed for integration testing)

**Pull request merged**